File: 1849_splitting_a_string_into_descending_consecutive_values/solution.py
# LeetCode 1849 - Splitting a String Into Descending Consecutive Values
# https://leetcode.com/problems/splitting-a-string-into-descending-consecutive-values/

class Solution:
    def splitString(self, s: str) -> bool:
        n = len(s)

        def dfs(index: int, previous: int | None, parts: int) -> bool:
            if index == n:
                return parts >= 2

            for end in range(index + 1, n + 1):
                value = int(s[index:end])
                if previous is None:
                    if dfs(end, value, parts + 1):
                        return True
                elif value == previous - 1:
                    if dfs(end, value, parts + 1):
                        return True
                elif value > previous - 1:
                    break

            return False

        return dfs(0, None, 0)

File: 1849_splitting_a_string_into_descending_consecutive_values/Solution.scala
// LeetCode 1849 - Splitting a String Into Descending Consecutive Values
// https://leetcode.com/problems/splitting-a-string-into-descending-consecutive-values/

object Solution {
  def splitString(s: String): Boolean = {
    val n = s.length

    def dfs(index: Int, previous: Option[BigInt], parts: Int): Boolean = {
      if (index == n) return parts >= 2
      var end = index + 1
      while (end <= n) {
        val value = BigInt(s.substring(index, end))
        previous match {
          case None =>
            if (dfs(end, Some(value), parts + 1)) return true
          case Some(prev) =>
            if (value == prev - 1) {
              if (dfs(end, Some(value), parts + 1)) return true
            } else if (value > prev - 1) return false
        }
        end += 1
      }
      false
    }

    dfs(0, None, 0)
  }
}
